File: gain-insight-fx/backend/app/api/v1/endpoints/currencies.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, Query

from app.api import deps
from app.schemas.user import UserInDB
from app.schemas.currency import ExchangeRate, CurrencyTrend
from app.services.currency import get_current_rates, get_historical_rates, get_currency_trend_analysis

router = APIRouter()

@router.get("/rates/current", response_model=List[ExchangeRate])
async def read_current_rates(
    current_user: UserInDB = Depends(deps.get_current_active_user)
):
    """
    Get current exchange rates for all tracked currencies.
    """
    return await get_current_rates()

@router.get("/rates/historical", response_model=List[ExchangeRate])
async def read_historical_rates(
    currency_code: str,
    days: int = Query(30, ge=1, le=365),
    current_user: UserInDB = Depends(deps.get_current_active_user)
):
    """
    Get historical exchange rates for a specific currency.
    """
    return await get_historical_rates(currency_code, days)

@router.get("/trends", response_model=CurrencyTrend)
async def read_currency_trends(
    currency_code: str,
    days: int = Query(30, ge=1, le=365),
    current_user: UserInDB = Depends(deps.get_current_active_user)
):
    """
    Get trend analysis for a specific currency.
    """
    return await get_currency_trend_analysis(currency_code, days)

File: frontend/src/api/currencies.js
import apiClient from "./client";

/**
 * Get current exchange rates
 * @param {Object} params - Query parameters
 * @returns {Promise<Object>} - Current rates
 */
export const getCurrentRates = (params = {}) => {
  return apiClient.get("/currencies/rates/current", { params });
};

/**
 * Get historical exchange rates
 * @param {Object} params - Query parameters
 * @returns {Promise<Object>} - Historical rates
 */
export const getHistoricalRates = (params = {}) => {
  return apiClient.get("/currencies/rates/historical", { params });
};

/**
 * Get currency trends
 * @param {Object} params - Query parameters
 * @returns {Promise<Array>} - Currency trends
 */
export const getCurrencyTrends = (params = {}) => {
  return apiClient.get("/currencies/trends", { params });
};

/**
 * Alias for getCurrencyTrends to maintain backward compatibility
 * @param {string} currencyCode - Currency code
 * @param {number} days - Number of days
 * @returns {Promise<Array>} - Currency trends
 */
export const getTrends = (currencyCode, days = 30) => {
  return apiClient.get(
    `/currencies/trends?currency_code=${currencyCode}&days=${days}`
  );
};
